Add scripts that sign the module for Secure Boot

flodavid opened the pull request:
I created scripts to automate the signing and installation steps required on systems with Secure Boot enabled

- For DKMS: /sign_and_install.sh checks that files configured as MOK for DKMS exist and the MOK is enrolled. If it is not the case, then it creates and/or enroll them using either Ubuntu’s tool or `openssl` (or `dkms`) and `mokutil` commands.
Then it run the install script that should build and sign the module using standard DKMS procedure.
- For generic: /hid-xpadneo/sign_and_install_generic.sh applies the same checks, but on “/var/lib/shim-signed/mok” as MOK files location.
Then it build and install the module using make commands.
- /lib/mok.sh handles MOK checks, creation and enrollment.
- /hid-xpadneo/Makefile has been modified to support signing after building the module. It uses MOK located in “/var/lib/shim-signed/mok”.
- /Makefile is modified to better clean files on uninstall.

<!-- By submitting a contribution, you agree that it is accepted under the project's primary license,
     with the additional permission that it may also be distributed under GPL-2.0-only. -->

- [x] I agree that my contribution may be distributed under GPL-2.0-only (in addition to the project's primary license)

flodavid committed:
xpadneo, installer: Add scripts that sign the module for Secure Boot

- New mok.sh in lib. Handles MOK check, creation and enrolment
  - Uses `dkms generate_mok` if available, with a good version
  and the default directory is used to store the MOK
  - Uses `update-secureboot-policy` if it is available,
  with a good version, and we are using its directory to store the MOK
  - Fallbacks to `openssl` and `mokutil` otherwise
- New scripts to sign using DKMS or generic install
- New signing rules in Makefile

kakra reviewed (changes requested):
Mostly nitpicks. I appreciate your work but I have to admit that I did just a quick scan through your changes and didn't test it yet. That's because I don't run a secure boot system here. I'd need to setup a VM or container - probably with different distributions.

If that works, tho, I'm pretty confident I can use that as a template to make my bare metal system secureboot. So, thanks already here. :-)

flodavid commented on hid-xpadneo/sign_and_install_generic.sh:
```diff
@@ -66,12 +65,15 @@ function signing_ko () {
   fi
-  info "Cert Password: $mokutil_509key_pass"
+
   zstd -f --decompress "${ko_folder}/hid-xpadneo.ko.zst"
```
Section removed

kakra commented on Makefile:
```diff
@@ -54,20 +56,17 @@ install: build
-	echo "Skipping key creation"
-endif
-	@$(KSRC)/scripts/sign-file sha256 MOK.priv MOK.der hid-xpadneo.ko
+	@sudo update-secureboot-policy --new-key
```
The Makefile should not call sudo, especially it should not do so hidden, and it should not run by surprise (e.g., you may have unlocked sudo before, and then it would run without asking for passwords). This also makes no sense because `modprobe` requires sudo but doesn't use it here. I'd suggest to rather add a guard `[ "$(id -u)" = "0"  ] || { echo >&2 "need to run with sudo"; exit 1; }`.

There are exceptions to this rule but this one is obviously not it.

kakra commented on sign_and_install.sh:
```diff
@@ -14,28 +15,22 @@ then error "Please run as root"
 command -v openssl > /dev/null 2>&1 || ( warn "missing openssl, you will not be able to create a new MOK key")
 command -v mokutil > /dev/null 2>&1 || ( warn "missing mokutil, you will not be able to sign the module")
-ls /var/lib/shim-signed > /dev/null 2>&1 || ( warn "missing shim-signed, you will not be able to sign the module with dkms")
+ls /var/lib/shim-signed > /dev/null 2>&1 || ( error "missing shim-signed, unable to enrol MOK")
```
Don't use `ls` to check for presence of a file name. You're expecting either a readable file or a directory. Use `[ -d /var/lib/shim-signed ] || error "missing ..."` or if you want to support symlinks, use `[ -d /var/lib/shim-signed/. ] || ...`. See `man test` for more options. Also, is it necessary to spawn `( subshells )` here?

kakra commented on sign_and_install.sh:
```diff
@@ -47,68 +42,16 @@ function mokutil_setup () {
-    generate_certificate
-    import_trust
+    sudo update-secureboot-policy --new-key
+    sudo update-secureboot-policy --enroll-key
```
Same...

kakra commented on sign_and_install.sh:
```diff
@@ -157,12 +102,14 @@ then
 else
   if mokutil --sb-state | grep -q "SecureBoot enabled"
-  then if ! grep -q "hid_xpadneo" /proc/modules && info "hid_xpadneo is loaded"
+  then if ! grep -q "hid_xpadneo" /proc/modules
```
This will fail on systems which linked xpadneo statically into the kernel. Check for `[ -d /sys/module/hid_xpadneo ]`, it will tell you if the driver is present in the kernel. OTOH, if xpadneo is statically linked into the kernel, you won't have to try signing it in the first place. ;-)

kakra replied:
modinfo will return `(builtin)` on the filename line if it is builtin, e.g. I have btrfs built into the kernel:
```bash
# modinfo btrfs
name:           btrfs
filename:       (builtin)
...
```

But if a module is unknown, it will return an error:
```bash
# modinfo unknown || echo "exit code: $?"
modinfo: ERROR: Module unknown not found.
exit code: 1
```

flodavid replied:
Now that I think about it, I do not need to worry about that, since, as you said, there is no need to sign it if the module is statically linked. And I already think the signing & install when the module is loaded (as it should be if statically linked I guess)

kakra reviewed (changes requested):
Good cleanup but there are some issues.

kakra commented on Makefile:
```diff
@@ -54,20 +56,17 @@ install: build
 
 uninstall: VERSION
-	$(DKMS) remove "hid-xpadneo/$(shell cat VERSION)" --all
+	$(DKMS) remove "hid-xpadneo/$(shell cat VERSION)" --all || true
```
On another note: This `Makefile` is to be used by package managers. We should not mess with `dkms` here. The docs explicitly state that it's up to the package maintainer to handle dkms.

flodavid replied:
Because if the module has already been removed, we has this error and the other commands are not run:
```
dkms remove "hid-xpadneo/v0.9-230-g73fd8cd-dirty" --all
Error! The module/version combo: hid-xpadneo-v0.9-230-g73fd8cd-dirty is not located in the DKMS tree.
make: *** [Makefile:55: uninstall] Error 3
```
I think adding a warning would be redundant.

kakra commented on Makefile:
```diff
@@ -47,9 +56,17 @@ install: build
+	@$(KSRC)/scripts/sign-file sha256 $(SHIM_MOK_DIR)/MOK.priv $(SHIM_MOK_DIR)/MOK.der $(LIB_PREFIX)/hid-xpadneo.ko
+	modprobe hid-xpadneo
+
+sign-install: all install sign
```
No need to pull in `all` here, `install` already pulls `build`.

kakra commented on Makefile:
```diff
@@ -47,9 +56,17 @@ install: build
 	rm -f $(MODPROBE_CONFS:%=$(PREFIX)$(ETC_PREFIX)/modprobe.d/%)
 	rmdir --ignore-fail-on-non-empty -p $(PREFIX)$(ETC_PREFIX)/modprobe.d $(PREFIX)$(ETC_PREFIX)/udev/rules.d $(PREFIX)$(DOC_PREFIX)
+
+sign:
```
You need `$(LIB_PREFIX)/hid-xpadneo.ko` here.

kakra reviewed (commented):
I wonder if signing with the `Makefile` is even worth it. It should be handle by dkms. If you need this for development, maybe move it to `hid-xpadneo/Makefile` instead? That file also allows usage of sudo for developer convenience. I'm using that to rebuild and reload my module during development like this:

```bash
make -C hid-xpadneo reinstall
```

flodavid commented on sign_and_install.sh:
```diff
@@ -119,7 +62,9 @@ function install_xpadneo () {
   source "$(dirname "$0")/install.sh"
 
-  ls /lib/modules/"$(uname -r)"/updates/
+  if [ ! -f /lib/modules/"$(uname -r)"/updates/dkms/hid-xpadneo.ko.zst ] && [ ! -f /lib/modules/"$(uname -r)"/updates/dkms/hid-xpadneo.ko.xz ]  && [ ! -f /lib/modules/"$(uname -r)"/updates/dkms/hid-xpadneo.ko ]
```
Done

flodavid commented on sign_and_install.sh:
```diff
@@ -14,28 +15,22 @@ then error "Please run as root"
     else
       warn "info the key exist but is not trusted yet, reimporting it because it does not hurt. You missing a reboot?"
-      import_trust
+      sudo update-secureboot-policy --enroll-key
```
I created a mok.sh lib

kakra commented:
> I have a problem on my system though. I guess I have firmware 5.x since the module is not working without `uhid`

I'm not sure how we should properly handle that. It is expected that bluez loads uhid if it is needed. xpadneo is not responsible to do that, we are a HID user, not a HID provider. So I think you should report it to your distribution, maybe they should link uhid statically into the kernel, or add a dependency.

Until then, you could just run:
```bash
# always load uhid on boot (requires systemd)
echo "uhid" | sudo tee -a /etc/modules-load.d/99-local.conf

# OR: load uhid with xpadneo (untested)
echo "softdep hid-xpadneo pre: uhid" | sudo tee -a /etc/modprobe.d/99-xpadneo-local.conf

# ALTERNATIVE: load uhid with bluetooth (untested, probably preferred)
echo "softdep bluetooth pre: uhid" | sudo tee -a /etc/modprobe.d/99-bluetooth-uhid.conf
```

If uhid-related reports keep popping up, we might need to add the latter into xpadneo, tho, I still think, distributions should add that to the standard kernel if they provide bluez with BLE support (or make the bluez service load the module). BLE input devices cannot be used without uhid. Thus, there's no point in adding a softdep to each driver which in theory might maybe support a Bluetooth HID device...

**EDIT:** I added another alternative.

kakra commented:
> What's strange is that I have no vibration when the controller is connected, even though there is `ff_connect_notify=1`

This can happen if xpadneo needs to bind lazily to the device via udev rules - it may not have worked during the first connection, and then only on second connect, it will work (and that means, also `uhid` is present, because we need that to speak to the controller). If xpadneo is already loaded, it will bind immediately and automatically.

You should see a difference in dmesg for both tries - if you can still replicate it...

> The one who does is `softdep bluetooth` in "99-bluetooth-uhid.conf".

Which proves that the driver providing the HID bus has to load it - which is bluez. xpadneo is only a HID consumer. It doesn't even know about BT. Most likely, `bluez` user-space should probably trigger loading `uhid` as soon as an BLE input device connects, loading just a `bluetooth` module isn't technically the correct trigger to load `uhid`.

If you asked me, `uhid` should be built statically into the kernel by your distribution. It's a needed core driver, not optional.

flodavid commented on sign_and_install.sh:
```diff
@@ -162,7 +165,12 @@ function install_xpadneo () {
   info "finished install_xpadneo"
-  lsmod | grep -q hid_xpadneo && info "hid_xpadneo loaded!"
+
+  if lsmod | grep -q hid_xpadneo
```
I now check for modprobe errors and use `[ -d /sys/module/hid_xpadneo ]` instead

flodavid commented:
I rebased on master and fixed the conflict.
I also fixed styling errors raised by the workflow. There are some left, but it is either on files I did not modify, or bash scripts that are flagged as missing a  SPDX-License-Identifier, but other bash files do not have it as well.

flodavid committed:
xpadneo, generic-makefile: Bring generic makefile up to par with DKMS one

- install step adds udev rules, conf, metainfo and docs install
- uninstall step removes files created by install plus the module file

flodavid commented:
I re-tested signing using the generic install and the  signing is still not made during the compilation by the kernel, but manually signing the module after its compilation is working well.

I added a second commit, that adds rules to the generic Makefile that installs/uninstalls the same files as the DKMS Makefile (docs, udev rules, ...). Tell me if you prefer not to have this or to make a separate PR.

Example output for generic install with signing (MOK was already created):
```shell
$ sudo ./sign_and_install_generic.sh
[info] Secure boot is enabled, signing is needed
[info] MOK Key /var/lib/shim-signed/mok/MOK.der found, already trusted.
[info] start install_xpadneo
make -C .. VERSION
make[1]: Entering directory '/home/flodavid/Dev/xpadneo'
{ [ -n "" ] && echo  || git describe --tags --dirty; } >VERSION
make[1]: Leaving directory '/home/flodavid/Dev/xpadneo'
make -C /lib/modules/6.17.0-23-generic/build INSTALL_MOD_DIR="kernel/drivers/hid" LD=ld.bfd M=/home/flodavid/Dev/xpadneo/hid-xpadneo/src VERSION="v0.11-pre-44-g9b1dbf5" modules
make[1]: Entering directory '/usr/src/linux-headers-6.17.0-23-generic'
make[2]: Entering directory '/home/flodavid/Dev/xpadneo/hid-xpadneo/src'
warning: the compiler differs from the one used to build the kernel
  The kernel was built by: x86_64-linux-gnu-gcc-13 (Ubuntu 13.3.0-6ubuntu2~24.04.1) 13.3.0
  You are using:           gcc-13 (Ubuntu 13.3.0-6ubuntu2~24.04.1) 13.3.0
  CC [M]  xpadneo/consumer.o
  CC [M]  xpadneo/core.o
  CC [M]  xpadneo/debug.o
  CC [M]  xpadneo/device.o
  CC [M]  xpadneo/events.o
  CC [M]  xpadneo/keyboard.o
  CC [M]  xpadneo/mappings.o
  CC [M]  xpadneo/mouse.o
  CC [M]  xpadneo/power.o
  CC [M]  xpadneo/quirks.o
  CC [M]  xpadneo/rumble.o
  CC [M]  xpadneo/synthetic.o
  LD [M]  hid-xpadneo.o
  MODPOST Module.symvers
  LD [M]  hid-xpadneo.ko
  BTF [M] hid-xpadneo.ko
Skipping BTF generation for hid-xpadneo.ko due to unavailability of vmlinux
make[2]: Leaving directory '/home/flodavid/Dev/xpadneo/hid-xpadneo/src'
make[1]: Leaving directory '/usr/src/linux-headers-6.17.0-23-generic'
[ "0" = "0"  ] || { echo >&2 "needs to be run with sudo"; exit 1; }
make -C /lib/modules/6.17.0-23-generic/build INSTALL_MOD_DIR="kernel/drivers/hid" LD=ld.bfd M=/home/flodavid/Dev/xpadneo/hid-xpadneo/src VERSION="v0.11-pre-44-g9b1dbf5" modules_install
make[1]: Entering directory '/usr/src/linux-headers-6.17.0-23-generic'
make[2]: Entering directory '/home/flodavid/Dev/xpadneo/hid-xpadneo/src'
  INSTALL /lib/modules/6.17.0-23-generic/kernel/drivers/hid/hid-xpadneo.ko
  SIGN    /lib/modules/6.17.0-23-generic/kernel/drivers/hid/hid-xpadneo.ko
  DEPMOD  /lib/modules/6.17.0-23-generic
Warning: modules_install: missing 'System.map' file. Skipping depmod.
make[2]: Leaving directory '/home/flodavid/Dev/xpadneo/hid-xpadneo/src'
make[1]: Leaving directory '/usr/src/linux-headers-6.17.0-23-generic'
mkdir -p /etc/modprobe.d /etc/udev/rules.d /usr/share/doc/xpadneo
install -D -m 0644 -t /etc/modprobe.d etc-modprobe.d/xpadneo.conf
install -D -m 0644 -t /etc/udev/rules.d etc-udev-rules.d/60-xpadneo.rules etc-udev-rules.d/70-xpadneo-disable-hidraw.rules
install -D -m 0644 -t /usr/share/doc/xpadneo ../NEWS.md ../docs/3P-BUGS.md ../docs/BT_DONGLES.md ../docs/CONFIGURATION.md ../docs/DEBUGGING.md ../docs/FIRMWARE_FLASHING.md ../docs/PACKAGING.md ../docs/README.md ../docs/SDL.md ../docs/TESTING.md ../docs/TROUBLESHOOTING.md
udevadm control --reload
install -D -m 0644 -t /usr/share/metainfo ../io.github.atar_axis.xpadneo.metainfo.xml
make check_sudo
make[1]: Entering directory '/home/flodavid/Dev/xpadneo/hid-xpadneo'
[ "0" = "0"  ] || { echo >&2 "needs to be run with sudo"; exit 1; }
make[1]: Leaving directory '/home/flodavid/Dev/xpadneo/hid-xpadneo'
update-secureboot-policy --enroll-key | grep -q --invert-match 'No DKMS modules installed' \
  || mokutil --import "/var/lib/shim-signed/mok/MOK.der"
Nothing to do.
SKIP: /var/lib/shim-signed/mok/MOK.der is already enrolled
/lib/modules/6.17.0-23-generic/build/scripts/sign-file sha256 /var/lib/shim-signed/mok/MOK.priv /var/lib/shim-signed/mok/MOK.der /lib/modules/6.17.0-23-generic/kernel/drivers/hid/hid-xpadneo.ko
rmmod hid-xpadneo || echo "proceeding anyways"
rmmod: ERROR: Module hid_xpadneo is not currently loaded
proceeding anyways
depmod -a
modprobe hid-xpadneo 
rm ../VERSION
filename:       /lib/modules/6.17.0-23-generic/kernel/drivers/hid/hid-xpadneo.ko
softdep:        pre: uhid
version:        v0.11-pre-44-g9b1dbf5
description:    Linux kernel driver for Xbox ONE S+ gamepads (Bluetooth), including rumble
author:         Kai Krakow <kai@kaishome.de>
author:         Florian Dollinger <dollinger.florian@gmx.de>
license:        GPL
srcversion:     FF4D64EAE1FAFD708AF1F63
alias:          hid:b0005g*v0000045Ep00000B13
alias:          hid:b0005g*v0000045Ep00000B22
alias:          hid:b0005g*v0000045Ep00000B05
alias:          hid:b0005g*v0000045Ep00000B20
alias:          hid:b0005g*v0000045Ep000002FD
alias:          hid:b0005g*v0000045Ep000002E0
depends:        hid
name:           hid_xpadneo
retpoline:      Y
vermagic:       6.17.0-23-generic SMP preempt mod_unload modversions 
sig_id:         PKCS#7
signer:         REDACTED Secure Boot signature key (xpadneo)"
sig_key:        REDACTED
sig_hashalgo:   sha256
signature:      REDACTED
parm:           trigger_rumble_mode:(u8) Trigger rumble mode. 0: pressure, 2: disable. (byte)
parm:           rumble_attenuation:(u8) Attenuate the rumble strength: all[,triggers] 0 (none, full rumble) to 100 (max, no rumble). (array of byte)
parm:           ff_connect_notify:(bool) Connection notification using force feedback. 1: enable, 0: disable. (bool)
parm:           quirks:(string) Override or change device quirks, specify as: "MAC1{:,+,-}quirks1[,...16]", MAC format = 11:22:33:44:55:66, no pulse parameters = 1, no trigger rumble = 2, no motor masking = 4, use Linux button mappings = 16, use Nintendo mappings = 32, use Share button mappings = 64, reversed motor masking = 128, swapped motor masking = 256, apply no heuristics = 512 (array of charp)
parm:           disable_mouse:(bool) Disable mouse device permanently. 0: allow mouse, 1: disallow mouse. (bool)
parm:           gamepad_compliance:(bool) Adhere to Linux Gamepad Specification by using signed axis values. 1: enable, 0: disable. (bool)
parm:           disable_deadzones:(bool) Disable dead zone handling for raw processing by Wine/Proton, confuses joydev. 0: disable, 1: enable. (bool)
parm:           disable_shift_mode:(bool) Disable use Xbox logo button as shift. Will prohibit profile switching when enabled. 0: disable, 1: enable. (bool)
parm:           debug_descriptor:(bool) Dump unpatched HID descriptor to dmesg. 0: auto, 1: always. (bool)
parm:           debug_hid:(bool) Debug HID reports. 0: disable, 1: enable. (bool)
[info] finished install_xpadneo
[info] hid_xpadneo loaded!
[info] finished install_xpadneo
```
